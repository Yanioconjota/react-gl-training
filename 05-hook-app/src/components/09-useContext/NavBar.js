import React from 'react';
import { Link, NavLink } from "react-router-dom";

const routes = [
  { path: "/", name: "Home" },
  { path: "/login", name: "Login" },
  { path: "/about", name: "About" },
];

export const NavBar = () => {
  return (
    <nav className="navbar navbar-expand-sm">
      <span className="navbar-brand">
        <Link className="nav-link" to="/">
          Use Context App
        </Link>
      </span>
      <div className="collapse navbar-collapse" id="navbarNav">
        <ul className="navbar-nav">
          {routes.map((route) => (
            <li className="nav-item" key={route.name}>
              <NavLink
                activeclassname="active"
                className="nav-link" to={route.path}>
                {route.name}
              </NavLink>
            </li>
          ))}
        </ul>
      </div>
    </nav>
  );
};