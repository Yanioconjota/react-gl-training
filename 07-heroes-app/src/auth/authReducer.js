import { types } from "../types/types";

//ejemplo de state
/* const state = {
  name: 'Homero',
  logged: true
}; */

//ejemplo de action
/* const loginAction = {
  type: types.login,
  payload: {
    name: 'Homero',
    email: 'homero@email.com'
  }
}; */

export const authReducer = (state = {}, action) => {
  switch (action.type) {
    case types.login:
      return {
        ...action.payload,
        logged: true
      }
    
    case types.logout:
      return {
        logged: false
      }
  
    default:
      return state;
  }
}