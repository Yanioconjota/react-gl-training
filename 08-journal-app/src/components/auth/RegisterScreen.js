import React from 'react';
import { Link } from "react-router-dom";
import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {

  const [formValues, handleInputChange] = useForm({
    name: 'Homero',
    email: "homerothompson@gmail.com",
    password: "123456",
    password2: "123456"
  });

  const { name, email, password, password2 } = formValues;

  const handleRegister = (e) => {
    e.preventDefault();
    console.log(name, email, password, password2);
  }

  return (
    <>
      <h3 className="auth__title">Register</h3>
      <form onSubmit={ handleRegister }>
        <input type="text"
               placeholder="Name"
               name="name"
               value={ name }
               onChange={ handleInputChange }
               className="auth__input"
               autoComplete="off"/>

        <input type="text"
               placeholder="Email"
               name="email"
               value={ email }
               onChange={ handleInputChange }
               className="auth__input"
               autoComplete="off"/>

        <input type="password" 
               placeholder="Password"
               name="password"
               value={ password }
               onChange={ handleInputChange }
               className="auth__input"/>

        <input type="password" 
               placeholder="Confirm password"
               name="password2"
               value={ password2 }
               onChange={ handleInputChange }
               className="auth__input"/>

        <button type="submit"
                className="btn btn-primary btn-block mb-5">Login
        </button>
        <Link className="link" to="/auth/login">
          Already registered?
        </Link>
      </form>
    </>
  )
}
