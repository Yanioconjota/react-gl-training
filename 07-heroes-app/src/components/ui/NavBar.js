import React, { useContext } from "react";
import { Link, NavLink, useNavigate } from "react-router-dom";
import { AuthContext } from "../../auth/authContext";
import { types } from "../../types/types";

export const Navbar = () => {

  const { user, dispatch } = useContext(AuthContext);
  
  const { name } = user;

  const navigate = useNavigate();

  const handleLogout = () => {
    dispatch({type: types.logout});
    //como 2 argumento recibe opciones, donde replace reemplaza esta ruta para no poder volver con el go back del browser
    navigate("/login", {
      replace: true,
    });
  }

  return (
    <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
      <div className="container-fluid">
        <Link className="navbar-brand" to="/">
          <strong>HEROAPP</strong>
        </Link>
        <div className="navbar-collapse">
          <div className="navbar-nav">
            <NavLink
              activeclassname="active"
              className="nav-item nav-link"
              to="/marvel">
              Marvel
            </NavLink>
            <NavLink
              activeclassname="active"
              className="nav-item nav-link"
              to="/dc">
              DC
            </NavLink>
            <NavLink
              activeclassname="active"
              className="nav-item nav-link"
              to="/all">
              Todos
            </NavLink>
            <NavLink
              activeclassname="active"
              className="nav-item nav-link"
              to="/search">
              Buscar
            </NavLink>
          </div>
        </div>
        <div className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end">
          <ul className="navbar-nav ml-auto">
            <li className="nav-item nav-link text-info">
              <span>{name}</span>
            </li>
            <li id="logout" className="nav-item nav-link"
              onClick={handleLogout}>
              Logout
            </li>
          </ul>
        </div>
      </div>
    </nav>
  );
};
