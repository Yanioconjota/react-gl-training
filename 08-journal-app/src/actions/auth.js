import Swal from "sweetalert2";
import { types } from "../types/types";
import {
  googleAuthProvider,
  getAuth,
  signInWithPopup,
  signInWithEmailAndPassword,
  createUserWithEmailAndPassword,
  signOut,
  updateProfile,
} from "../firebase/firebase-config";
import { finishLoading, startLoading } from "./ui";
import { notesLogout } from "./notes";

//gracias al midleware thunk-redux tenemos acceso al dispatch dentro del return
export const startLoginWithEmailPassword = (email, password) => {
  return (dispatch) => {
    const auth = getAuth();
    dispatch(startLoading());
    signInWithEmailAndPassword(auth, email, password)
      .then(({ user }) => {
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch((err) => {
        dispatch(finishLoading());
        //mensaje de error en modal
        Swal.fire('Error', err.message, 'error');
      });
  };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
  return (dispatch) => {
    const auth = getAuth();
    dispatch(startLoading());
    createUserWithEmailAndPassword(auth, email, password)
      .then(async ({ user }) => {
        //update profile agrega el displayName, ya que este solo se crea cuando nos registramos con google o cualquier red social
        await updateProfile(user, { displayName: name });
        dispatch(login(user.uid, user.displayName));
        dispatch(finishLoading());
      })
      .catch((err) => {
        dispatch(finishLoading());
        //mensaje de error en modal
        Swal.fire("Error", err.message, "error");
      });
  };
};

export const startGoogleLogin = () => {
  return (dispatch) => {
    const auth = getAuth();
    signInWithPopup(auth, googleAuthProvider).then(({ user }) => {
      dispatch(login(user.uid, user.displayName));
    });
  };
};

export const login = (uid, displayName) => ({
  type: types.login,
  payload: {
    uid,
    displayName,
  },
});

export const startLogout = () => {
  return async (dispatch) => {
    const auth = getAuth();
    await signOut(auth);
    dispatch(logout());
    dispatch(notesLogout());
  };
};

export const logout = () => ({
  type: types.logout,
});
