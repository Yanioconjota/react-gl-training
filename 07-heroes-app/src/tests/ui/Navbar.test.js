import "@testing-library/jest-dom";
import { mount } from "enzyme";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { AuthContext } from "../../auth/authContext";
import { Navbar } from "../../components/ui/NavBar";

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe('Pruebas con Navbar', () => {

  const typesDemo = {
    login: "[auth] Login",
    logout: "[auth] Logout",
  };

  const contextValue = {
    user: {
      logged: true,
      name: "Homero",
    },
    dispatch: jest.fn(),
  };

  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      {/* Permite hacer pruebas en componentes que hacen uso del useNavigate */}
      <MemoryRouter initialEntries={["/"]}>
        <Routes>
          <Route path="/" element={<Navbar/>}/>
        </Routes>
      </MemoryRouter>
    </AuthContext.Provider>
  );
  test('debe mostrar el componente correctamente', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".text-info > span").text().trim()).toBe("Homero");
  });

  test('debe llamar el logout, el navigate y el dispatch con los parÃ¡metros correctos', () => {
    wrapper.find('#logout').simulate('click');
    //expect(handleLogout).toHaveBeenCalled();
    expect(contextValue.dispatch).toHaveBeenCalledWith({
      type: typesDemo.logout,
    });
    expect(mockNavigate).toHaveBeenCalledWith('/login', {replace: true});
  });
});