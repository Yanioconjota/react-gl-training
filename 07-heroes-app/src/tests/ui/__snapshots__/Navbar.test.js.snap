// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con Navbar debe mostrar el componente correctamente 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark"
>
  <div
    className="container-fluid"
  >
    <a
      className="navbar-brand"
      href="/"
      onClick={[Function]}
    >
      <strong>
        HEROAPP
      </strong>
    </a>
    <div
      className="navbar-collapse"
    >
      <div
        className="navbar-nav"
      >
        <a
          activeclassname="active"
          className="nav-item nav-link"
          href="/marvel"
          onClick={[Function]}
        >
          Marvel
        </a>
        <a
          activeclassname="active"
          className="nav-item nav-link"
          href="/dc"
          onClick={[Function]}
        >
          DC
        </a>
        <a
          activeclassname="active"
          className="nav-item nav-link"
          href="/all"
          onClick={[Function]}
        >
          Todos
        </a>
        <a
          activeclassname="active"
          className="nav-item nav-link"
          href="/search"
          onClick={[Function]}
        >
          Buscar
        </a>
      </div>
    </div>
    <div
      className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
    >
      <ul
        className="navbar-nav ml-auto"
      >
        <li
          className="nav-item nav-link text-info"
        >
          <span>
            Homero
          </span>
        </li>
        <li
          className="nav-item nav-link"
          id="logout"
          onClick={[Function]}
        >
          Logout
        </li>
      </ul>
    </div>
  </div>
</nav>
`;
